ssl/s3_lib.c

```diff
@@ -3770,6 +3769,34 @@ OPENSSL_GLOBAL const SSL_CIPHER ssl3_ciphers[] = {
      256,
      },
 
+    {
+     1,
+     "GOST2012-GOST8912-GOST8912",
+     0x0300ff85,
+     SSL_kGOST,
+     SSL_aGOST12 | SSL_aGOST01,
+     SSL_eGOST2814789CNT12,
+     SSL_GOST89MAC12,
+     SSL_TLSV1,
+     SSL_NOT_EXP | SSL_HIGH,
+     SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC,
```

Comment:
Ahhh. Ok. I think I understand now. So the acceptable hash algorithms in the cert verify could be any of the GOST hashes. However the handshake hash is always a specifically defined one. Am I correct?

If so, I think a possible answer is to build on the SIGALGS code. Normally we calculate all the digests that we're going to need for the Finished message as we go and discard the handshake messages once we've processed them (so there is no opportunity to go back and calculate a different digest). With SIGALGS though if the SIGALGS extension is present we retain all the handshake messages received so far in `s->s3->handshake_buffer`. Only after the CertVerify do we get rid of that and start calculating the digests as we go. So in outline I think the changes you would have to make would be to:

* Revert the above line back to:
````
SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC,
````
* In s3_srvr.c in function ssl3_accept in the SSL3_ST_SR_KEY_EXCH_A/B state where there is a test for SIGALGS:
````
            } else if (SSL_USE_SIGALGS(s)) {
                s->state = SSL3_ST_SR_CERT_VRFY_A;
                s->init_num = 0;
                if (!s->session->peer)
                    break;
                if (!s->s3->handshake_buffer) {
                    SSLerr(SSL_F_SSL3_ACCEPT, ERR_R_INTERNAL_ERROR);
                    s->state = SSL_ST_ERR;
                    return -1;
                }
                /*
                 * For sigalgs freeze the handshake buffer. If we support
                 * extms we've done this already so this is a no-op
                 */
                if (!ssl3_digest_cached_records(s, 1)) {
                    s->state = SSL_ST_ERR;
                    return -1;
                }
````
We additionally want to take that branch if we are using a GOST algorithm so add an OR clause to the if statement. So what we are doing here is freezing the state of s->s3->handshake_buffer ready for later use by the cert verify code, and initialising the digests needed to calculate the Finished mac.
* In ssl3_get_cert_verify we need to change GOST code to work in a similar way to the SIGALGS code just a bit above it, i.e. this code:
````
    if (SSL_USE_SIGALGS(s)) {
        long hdatalen = 0;
        void *hdata;
        hdatalen = BIO_get_mem_data(s->s3->handshake_buffer, &hdata);
        if (hdatalen <= 0) {
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, ERR_R_INTERNAL_ERROR);
            al = SSL_AD_INTERNAL_ERROR;
            goto f_err;
        }
#ifdef SSL_DEBUG
        fprintf(stderr, "Using TLS 1.2 with client verify alg %s\n",
                EVP_MD_name(md));
#endif
        if (!EVP_VerifyInit_ex(&mctx, md, NULL)
            || !EVP_VerifyUpdate(&mctx, hdata, hdatalen)) {
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, ERR_R_EVP_LIB);
            al = SSL_AD_INTERNAL_ERROR;
            goto f_err;
        }

        if (EVP_VerifyFinal(&mctx, data, len, pkey) <= 0) {
            al = SSL_AD_DECRYPT_ERROR;
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, SSL_R_BAD_SIGNATURE);
            goto f_err;
        }
    }
````
I hope I've understood correctly and that makes some kind of sense :-)